### ISSUE TITLE
Abnormal training behavior with Yolov8?

### ISSUE THREAD

**Issue by paolocomis** [ultralytics/ultralytics#1044]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

After a decent experience using Yolov5, I decided to test the newer Yolov8 to see the real differences in terms of detection accuracy.

Premise:
- Same data set used (20K of images for 2 classes)
- Same model version "L" version
- Same number of training epochs 600

I immediately noticed a different approach:
1: After the first epoch map50 and map50-95 showed a very high value (0.838 - 0.524) compared to the first epoch with yolov5 (0.054 - 0.025).
2: Yolov8's training (training in progress) seems to have peaked at its highest accuracy after only 100 epochs.
3: Unlike Yolov5, the `box_loss`  `cls_loss`  `dfl_loss` values remain more or less unchanged during the eras with perhaps high values, and do not gradually decrease.

The command executed is as follows:
`yolo task=detect device=0 mode=train model=yolov8l.pt cache=ram batch=16 workers=48 data=custom_data.yaml epochs=600 imgsz=1280 name=my_project plots=True exist_ok=True`

I attach some pictures of the training with Yolov8 currently running:

![metrica](https://user-images.githubusercontent.com/18678028/219953098-70c97f6a-aa58-4441-83e6-d37921da7a5a.png)
![train](https://user-images.githubusercontent.com/18678028/219953105-bf100452-94d4-482d-85be-88facb648b2d.png)
![val](https://user-images.githubusercontent.com/18678028/219953124-3650ba6a-2ab5-47b4-95e2-a2ad3ae5ee94.png)

Am I doing something wrong?

**Comment by Legor** [comment]
Hey, how does it compare to your yolov5 training, especially regarding your mentioned map vals or the losses?
The fact, that the highest map values are achieved earlier and that the lowest loss values are reached earlier as well doesn't indicate that something is going wrong, quite the opposite, actually. If however your yolov5 training achieved much better results, that would be something different.

**Comment by paolocomis** [comment]
> Ehi, come si confronta con il tuo allenamento yolov5, in particolare per quanto riguarda i valori della mappa citati o le perdite? Il fatto che i valori di mappa più alti sono raggiunti prima e che anche i valori di perdita più bassi sono raggiunti prima non indica che qualcosa stia andando storto, anzi, anzi. Se tuttavia il tuo allenamento yolov5 ha ottenuto risultati molto migliori, sarebbe qualcosa di diverso.

Thanks for the reply. Maybe I was unclear, sorry. Yolov5 didn't get better results, indeed, Yolov8 is getting a map0.5-0.9 value already after only 100 epochs.
I was referring mostly to the values of the losses. For example the "box loss" value: at the end of the training with Yolov5 its value was around 0.2, instead after 100 training periods with Yolov8 the value seems to have stabilized around the value 0.8 and from the curve in the attached image there seems to be scope for further lowering. So I was wondering if this behavior with Yolov8 is normal or maybe it's better to change some settings.

**Comment by Legor** [comment]
If the data and the loss function are the same used as in yolov5, It can be expected that they get to a similar value, given enough training time. The improvements in the later epochs are usually very small, so it may be misleading to judge by looking at the plots only. It may well happen, that for the next 100 epochs you still get marginal improvements. From my personal experience, it's often better to set the target epoch value to a larger value like 400 and let the training be terminated by the early stopping mechanisms (see parameter 'patience')

**Comment by rgaufman** [comment]
I have something much worse and wondering if I'm missing something obvious. I'm running:

```
$ yolo task=detect mode=train model=yolov8l.pt data=/home/hackeron/Trash-Detection-2/data.yaml epochs=300 batch=12
New https://pypi.org/project/ultralytics/8.0.102 available 😃 Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.0.101 🚀 Python-3.10.6 torch-2.0.1+cu117 CUDA:0 (NVIDIA GeForce RTX 3060 Ti, 8191MiB)
yolo/engine/trainer: task=detect, mode=train, model=yolov8l.pt, data=/home/hackeron/Trash-Detection-2/data.yaml, epochs=300, patience=50, batch=12, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=None, exist_ok=False, pretrained=False, optimizer=SGD, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=0, resume=False, amp=True, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, show=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, vid_stride=1, line_width=None, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, cfg=None, v5loader=False, tracker=botsort.yaml, save_dir=runs/detect/train8
```
I'm about 130 Epochs in and the results I'm seeing are this so far. The mAP(50) is all over the place and seems to be flattening out at around 0.13 which seems extremely low. Any ideas what could be causing this?

![Screenshot 2023-05-16 at 01 12 41](https://github.com/ultralytics/ultralytics/assets/33842342/6971a858-30bc-4711-9b09-cabac9a0e205)

**Comment by glenn-jocher** [comment]
@rgaufman hey there! 

Firstly, thank you for using YOLOv8 and for reaching out to us. 

Regarding the `mAP50` fluctuation and eventual flattening out, this could potentially be caused by a number of factors such as:

1. Insufficient training data, especially if the dataset is smaller or less diverse than what is generally used with YOLOv8. It's recommended to have at least several thousand images per class.

2. Incorrect `data.yaml` settings such as incorrect paths to images or labels.

3. Incorrect hyperparameters in the `yolo` command, which could adversely affect the learning rate and other optimization settings.

4. High imbalance in the number of objects per class in the dataset being used, which can cause difficulty in training the model.

We suggest reviewing your `data.yaml` file and other training settings to ensure that they are optimal and suitable for the dataset you are using. Additionally, you may also experiment with adjusting hyperparameters such as the learning rate or the `box_loss`, `cls_loss`, and `dfl_loss` values to see if there is an improvement in the results.

Please let us know if this helps or if you have any further questions or concerns - we're here to help!

**Comment by cisco-silva** [comment]
@glenn-jocher Hi!

When I train the YOLOv8 model on my GPU, I have noticed that the mAP metrics consistently decrease (loss metrics increase) or become erratic instead of improving after each epoch. This is a perplexing issue since running the same code on my CPU yields the expected results, with mAP and loss improving as the number of epochs increases. 

**I am using the following configuration:**
Ubuntu 20.04
Python 3.11.5
NVIDIA GTX 1650
NVIDIA Driver 525.125.06
CUDA 11.8
cuDNN 8.6
Torch 2.1.0+cu118
Ultralytics YOLOv8.0.196

**Please see below the code used for traning:**

```
from ultralytics import YOLO
model = YOLO('yolov8n.pt')
results = model.train(data='coco128.yaml', epochs=60, device=0)
```
**Observation:**

I have included below two images that illustrate the problem I am facing:

1. Shows the model metrics after 54 epochs in the GPU. The model was supposed to do 60 epochs but it stopped at epoch 54 saying that it did not observe any improvement in the last 50 epochs, and selected the results at epoch 4 as the best.pt 
2. Shows the model metrics after 60 epochs in the CPU

**1 -Results with GPU**
![results_GPU](https://github.com/ultralytics/ultralytics/assets/147659342/8df7150c-9e60-44f4-a6a0-5533f18139f5)

**2 -Results with CPU (same code except for `device=cpu`)**
![results_CPU](https://github.com/ultralytics/ultralytics/assets/147659342/d9623a56-60ee-44f0-b21e-d53781662a03)

I have already tried various troubleshooting steps, including checking my GPU drivers, ensuring the correct CUDA and cuDNN versions, but the problem persists

**Comment by glenn-jocher** [comment]
@cisco-silva hi there,

Thank you for using YOLOv8 and for providing a detailed explanation of the issue you're encountering.

It is unusual for the behavior of the model to vary so significantly depending on the device it's running on, with the same dataset and code. From the images you've provided, it's clear that the GPU results are not as expected, and the training seems to be stuck and not converging optimally.

From the GPU metrics, it appears that the mAP is decreasing and loss is increasing, suggesting that the model is not learning effectively. This could potentially be due to a number of reasons, such as:

1. GPU-specific issues: There might be some GPU-specific settings or configurations that may affect how CUDA and cuDNN operate, leading to differences in training results.

2. Memory issues: Training deep learning models on GPUs requires ample memory. If insufficient memory is allocated for your model, it may lead to suboptimal training performance.

3. Variable precision: GPUs often run computations in float32 or lower precision. If this is impacting the learning of the model, toggling the 'amp' parameter on and off could potentially help identify if this is contributing to your issue.

Also, the inconsistencies between CPU and GPU could be due to underlying hardware or driver issues. It may be worth exploring if the GPU is performing correctly using other software or benchmarks, just to rule out this as a potential problem.

Kindly verify these aspects and let us know how it goes. We hope this helps you in analyzing the situation and finding a solution.
